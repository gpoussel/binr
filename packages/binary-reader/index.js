const BinaryReader = require("./lib/binary-reader");

module.exports = {
  BinaryReader,
};
