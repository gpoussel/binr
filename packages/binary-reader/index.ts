export * from "./lib/binary-reader";
