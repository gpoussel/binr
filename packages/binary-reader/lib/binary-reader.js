"use strict";

const _ = require("lodash");
const assert = require("assert");

const { StructureType } = require("@binr/definition-reader");
const BufferWrapper = require("./buffer-wrapper");

class BinaryReader {
  read(binaryBuffer, definition, providedStructureName = undefined) {
    assert(_.isObject(definition), "definition must be an object");
    assert(_.isBuffer(binaryBuffer), "binaryBuffer must be a buffer");

    let mainStructureName;
    if (_.isUndefined(providedStructureName)) {
      // No structure provided: assume there is only one in the definition
      // Otherwise that's an error
      assert(
        _.get(definition, "structures.length") === 1,
        `No 'structureName' provided: structures.length must be === 1`
      );
      mainStructureName = _.first(definition.structures).name;
    } else {
      mainStructureName = providedStructureName;
    }

    const mainStructure = _.find(definition.structures, s => s.name === mainStructureName);
    assert(!_.isUndefined(mainStructure), `Main structure '${mainStructureName} not found`);

    const structureType = new StructureType(mainStructure);
    const bufferWrapper = new BufferWrapper(binaryBuffer, mainStructure.getEndianness());
    return structureType.read(bufferWrapper);
  }
}

module.exports = BinaryReader;
