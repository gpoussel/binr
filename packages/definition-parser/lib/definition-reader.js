"use strict";

const _ = require("lodash");
const { DefinitionLexer } = require("./definition-lexer");
const DefinitionParser = require("./definition-parser");

class DefinitionReader {
  readInput(input) {
    if (!_.isString(input)) {
      throw new Error("input must be a string");
    }

    const lexingResult = new DefinitionLexer().tokenize(input);
    if (!_.isEmpty(lexingResult.errors)) {
      throw new Error(`Got an error while lexing input: ${_.first(lexingResult.errors).message}`);
    }

    const parser = new DefinitionParser();
    parser.input = lexingResult.tokens;
    const parsingResult = parser.definition();

    if (!_.isEmpty(parser.errors)) {
      throw new Error(`Got an error while parsing input: ${_.first(parser.errors).message}`);
    }

    class Visitor extends parser.getBaseCstVisitorConstructorWithDefaults() {
      constructor() {
        super();
        this.validateVisitor();
      }

      definition(ctx) {
        return {
          headers: _.map(ctx.headerClause, clause => this.visit(clause)),
          structures: _.map(ctx.structureClause, clause => this.visit(clause)),
        };
      }

      headerClause(ctx) {
        const name = this.getIdentifierName(_.first(ctx.IdentifierToken));
        const value = this.visit(_.first(ctx.valueClause));
        return {
          name,
          value,
        };
      }

      structureClause(ctx) {
        const exported = _.has(ctx, "ExportToken");
        const name = this.getIdentifierName(_.first(ctx.IdentifierToken));
        const fields = ctx.fieldClause.map(clause => this.visit(clause));
        return {
          name,
          exported,
          fields,
        };
      }

      fieldClause(ctx) {
        const type = this.getIdentifierName(_.get(ctx.IdentifierToken, 0));
        const typeRestriction = _.has(ctx, "ColonToken")
          ? this.getNumberValue(_.first(ctx.NumberLiteralToken))
          : undefined;
        const value = _.has(ctx, "EqualsToken") ? this.visit(_.first(ctx.valueClause)) : undefined;
        const name = this.getIdentifierName(_.get(ctx.IdentifierToken, 1));
        return {
          type,
          typeRestriction,
          name,
          value,
        };
      }

      valueClause(ctx) {
        if (_.has(ctx, "StringLiteralToken")) {
          return JSON.parse(_.first(ctx.StringLiteralToken).image);
        }
        throw new Error("Unhandled value type: ", _.keys(ctx));
      }

      getIdentifierName(identifier) {
        return identifier.image;
      }

      getNumberValue(literal) {
        return parseInt(literal.image, 10);
      }
    }

    const visitor = new Visitor();
    return visitor.visit(parsingResult);
  }
}

module.exports = DefinitionReader;
