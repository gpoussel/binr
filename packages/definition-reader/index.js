const { DefinitionReader } = require("./lib/definition-reader");
const { DefinitionBuilder } = require("./lib/definition-builder");
const { DefinitionLexer } = require("./lib/definition-lexer");
const { DefinitionParser } = require("./lib/definition-parser");
const { DefinitionValidator } = require("./lib/definition-validator");
const { StringType, UintType, IntType, StructureType } = require("./lib/types");

module.exports = {
  DefinitionReader,
  DefinitionBuilder,
  DefinitionLexer,
  DefinitionParser,
  DefinitionValidator,
  StringType,
  UintType,
  IntType,
  StructureType,
};
