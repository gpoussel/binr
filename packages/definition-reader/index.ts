export * from "./lib/definition-reader";
export * from "./lib/definition-builder";
export * from "./lib/definition-lexer";
export * from "./lib/definition-parser";
export * from "./lib/definition-validator";
export * from "./lib/types";
