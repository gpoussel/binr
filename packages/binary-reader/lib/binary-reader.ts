import * as assert from "assert";
import { find, get, isBuffer, isObject, isUndefined } from "lodash";

import { StructureType } from "@binr/definition-reader";
import { BufferWrapper, Environment, StreamObject } from "@binr/shared";

export class BinaryReader {
  public read(binaryBuffer, definition, providedStructureName?) {
    assert(isObject(definition), "definition must be an object");
    assert(isBuffer(binaryBuffer), "binaryBuffer must be a buffer");

    let mainStructureName;
    if (isUndefined(providedStructureName)) {
      // No structure provided: assume there is only one in the definition
      // Otherwise that's an error
      assert(
        get(definition, "structures.length") === 1,
        "No 'structureName' provided: structures.length must be === 1",
      );
      mainStructureName = definition.structures[0].name;
    } else {
      mainStructureName = providedStructureName;
    }

    const mainStructure = find(definition.structures, (s) => s.name === mainStructureName);
    assert(!isUndefined(mainStructure), `Main structure '${mainStructureName}' not found`);

    const structureType = new StructureType(mainStructure);
    const bufferWrapper = new BufferWrapper(binaryBuffer, mainStructure.endianness);

    const streamObject = new StreamObject(bufferWrapper);
    const environment = new Environment(streamObject);

    return structureType.read(bufferWrapper, environment);
  }
}
