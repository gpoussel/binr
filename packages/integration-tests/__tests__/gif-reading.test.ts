import * as fs from "fs";

import { BinaryReader } from "@binr/binary-reader";
import { DefinitionReader } from "@binr/definition-reader";

const pathToBinaryFixtures = `${__dirname}/../__fixtures__/binaries`;
const pathToDefinitionFixtures = `${__dirname}/../__fixtures__/definitions`;

describe("GIF reading", () => {
  const definitionReader = new DefinitionReader();

  const gifDefinitionFile = fs.readFileSync(`${pathToDefinitionFixtures}/gif.binr`);
  const gifDefinition = definitionReader.readInput(gifDefinitionFile.toString());

  const treeGifFile = `${pathToBinaryFixtures}/tree.gif`;
  const treeGifBuffer = fs.readFileSync(treeGifFile);

  test("reads GIF file", () => {
    const binaryReader = new BinaryReader();
    const gifValue = binaryReader.read(treeGifBuffer, gifDefinition, "GifFile");
    expect(gifValue).toBeDefined();
    expect(gifValue).toMatchSnapshot();
  });
});
