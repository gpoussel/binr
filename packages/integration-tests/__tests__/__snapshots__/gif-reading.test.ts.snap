// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIF reading reads GIF file 1`] = `
Object {
  "data": Object {},
  "globalColorTable": Array [
    Object {
      "blue": 1,
      "green": 128,
      "red": 1,
    },
    Object {
      "blue": 255,
      "green": 255,
      "red": 255,
    },
    Object {
      "blue": 1,
      "green": 65,
      "red": 128,
    },
    Object {
      "blue": 22,
      "green": 24,
      "red": 22,
    },
    Object {
      "blue": 6,
      "green": 123,
      "red": 7,
    },
    Object {
      "blue": 1,
      "green": 1,
      "red": 252,
    },
    Object {
      "blue": 16,
      "green": 128,
      "red": 15,
    },
    Object {
      "blue": 240,
      "green": 240,
      "red": 239,
    },
    Object {
      "blue": 245,
      "green": 167,
      "red": 5,
    },
    Object {
      "blue": 24,
      "green": 39,
      "red": 24,
    },
    Object {
      "blue": 4,
      "green": 251,
      "red": 128,
    },
    Object {
      "blue": 32,
      "green": 31,
      "red": 31,
    },
    Object {
      "blue": 38,
      "green": 38,
      "red": 38,
    },
    Object {
      "blue": 231,
      "green": 231,
      "red": 231,
    },
    Object {
      "blue": 216,
      "green": 216,
      "red": 216,
    },
    Object {
      "blue": 21,
      "green": 129,
      "red": 21,
    },
    Object {
      "blue": 32,
      "green": 120,
      "red": 31,
    },
    Object {
      "blue": 36,
      "green": 70,
      "red": 36,
    },
    Object {
      "blue": 4,
      "green": 201,
      "red": 251,
    },
    Object {
      "blue": 199,
      "green": 199,
      "red": 199,
    },
    Object {
      "blue": 38,
      "green": 88,
      "red": 38,
    },
    Object {
      "blue": 16,
      "green": 21,
      "red": 15,
    },
    Object {
      "blue": 23,
      "green": 120,
      "red": 23,
    },
    Object {
      "blue": 224,
      "green": 223,
      "red": 223,
    },
    Object {
      "blue": 35,
      "green": 55,
      "red": 35,
    },
    Object {
      "blue": 37,
      "green": 118,
      "red": 37,
    },
    Object {
      "blue": 38,
      "green": 104,
      "red": 38,
    },
    Object {
      "blue": 64,
      "green": 63,
      "red": 63,
    },
    Object {
      "blue": 33,
      "green": 56,
      "red": 29,
    },
    Object {
      "blue": 32,
      "green": 103,
      "red": 31,
    },
    Object {
      "blue": 192,
      "green": 191,
      "red": 191,
    },
    Object {
      "blue": 54,
      "green": 54,
      "red": 54,
    },
    Object {
      "blue": 32,
      "green": 37,
      "red": 31,
    },
    Object {
      "blue": 71,
      "green": 70,
      "red": 70,
    },
    Object {
      "blue": 167,
      "green": 167,
      "red": 167,
    },
    Object {
      "blue": 13,
      "green": 14,
      "red": 17,
    },
    Object {
      "blue": 183,
      "green": 183,
      "red": 183,
    },
    Object {
      "blue": 128,
      "green": 127,
      "red": 127,
    },
    Object {
      "blue": 151,
      "green": 151,
      "red": 151,
    },
    Object {
      "blue": 48,
      "green": 46,
      "red": 46,
    },
    Object {
      "blue": 26,
      "green": 55,
      "red": 26,
    },
    Object {
      "blue": 160,
      "green": 159,
      "red": 159,
    },
    Object {
      "blue": 10,
      "green": 10,
      "red": 9,
    },
    Object {
      "blue": 16,
      "green": 127,
      "red": 15,
    },
    Object {
      "blue": 25,
      "green": 104,
      "red": 25,
    },
    Object {
      "blue": 25,
      "green": 87,
      "red": 25,
    },
    Object {
      "blue": 176,
      "green": 175,
      "red": 175,
    },
    Object {
      "blue": 87,
      "green": 87,
      "red": 87,
    },
    Object {
      "blue": 36,
      "green": 129,
      "red": 36,
    },
    Object {
      "blue": 119,
      "green": 119,
      "red": 119,
    },
    Object {
      "blue": 80,
      "green": 79,
      "red": 79,
    },
    Object {
      "blue": 15,
      "green": 123,
      "red": 19,
    },
    Object {
      "blue": 33,
      "green": 71,
      "red": 30,
    },
    Object {
      "blue": 48,
      "green": 48,
      "red": 47,
    },
    Object {
      "blue": 7,
      "green": 105,
      "red": 7,
    },
    Object {
      "blue": 51,
      "green": 115,
      "red": 50,
    },
    Object {
      "blue": 103,
      "green": 103,
      "red": 102,
    },
    Object {
      "blue": 144,
      "green": 143,
      "red": 143,
    },
    Object {
      "blue": 135,
      "green": 135,
      "red": 135,
    },
    Object {
      "blue": 48,
      "green": 88,
      "red": 47,
    },
    Object {
      "blue": 1,
      "green": 33,
      "red": 66,
    },
    Object {
      "blue": 13,
      "green": 19,
      "red": 12,
    },
    Object {
      "blue": 96,
      "green": 95,
      "red": 95,
    },
    Object {
      "blue": 112,
      "green": 112,
      "red": 111,
    },
    Object {
      "blue": 52,
      "green": 100,
      "red": 51,
    },
    Object {
      "blue": 96,
      "green": 95,
      "red": 96,
    },
    Object {
      "blue": 9,
      "green": 207,
      "red": 108,
    },
    Object {
      "blue": 16,
      "green": 100,
      "red": 15,
    },
    Object {
      "blue": 19,
      "green": 74,
      "red": 128,
    },
    Object {
      "blue": 146,
      "green": 145,
      "red": 143,
    },
    Object {
      "blue": 112,
      "green": 110,
      "red": 108,
    },
    Object {
      "blue": 4,
      "green": 243,
      "red": 124,
    },
    Object {
      "blue": 48,
      "green": 71,
      "red": 47,
    },
    Object {
      "blue": 8,
      "green": 87,
      "red": 8,
    },
    Object {
      "blue": 24,
      "green": 35,
      "red": 39,
    },
    Object {
      "blue": 11,
      "green": 135,
      "red": 21,
    },
    Object {
      "blue": 3,
      "green": 3,
      "red": 231,
    },
    Object {
      "blue": 7,
      "green": 65,
      "red": 123,
    },
    Object {
      "blue": 13,
      "green": 84,
      "red": 18,
    },
    Object {
      "blue": 47,
      "green": 67,
      "red": 48,
    },
    Object {
      "blue": 25,
      "green": 72,
      "red": 26,
    },
    Object {
      "blue": 26,
      "green": 41,
      "red": 53,
    },
    Object {
      "blue": 18,
      "green": 122,
      "red": 67,
    },
    Object {
      "blue": 211,
      "green": 145,
      "red": 5,
    },
    Object {
      "blue": 20,
      "green": 28,
      "red": 35,
    },
    Object {
      "blue": 12,
      "green": 169,
      "red": 90,
    },
    Object {
      "blue": 16,
      "green": 92,
      "red": 15,
    },
    Object {
      "blue": 52,
      "green": 86,
      "red": 52,
    },
    Object {
      "blue": 83,
      "green": 82,
      "red": 77,
    },
    Object {
      "blue": 16,
      "green": 68,
      "red": 14,
    },
    Object {
      "blue": 14,
      "green": 69,
      "red": 20,
    },
    Object {
      "blue": 14,
      "green": 143,
      "red": 77,
    },
    Object {
      "blue": 23,
      "green": 142,
      "red": 46,
    },
    Object {
      "blue": 11,
      "green": 186,
      "red": 98,
    },
    Object {
      "blue": 228,
      "green": 155,
      "red": 3,
    },
    Object {
      "blue": 24,
      "green": 72,
      "red": 120,
    },
    Object {
      "blue": 253,
      "green": 128,
      "red": 253,
    },
    Object {
      "blue": 121,
      "green": 129,
      "red": 123,
    },
    Object {
      "blue": 147,
      "green": 106,
      "red": 11,
    },
    Object {
      "blue": 19,
      "green": 58,
      "red": 12,
    },
    Object {
      "blue": 19,
      "green": 196,
      "red": 241,
    },
    Object {
      "blue": 8,
      "green": 10,
      "red": 153,
    },
    Object {
      "blue": 13,
      "green": 17,
      "red": 22,
    },
    Object {
      "blue": 8,
      "green": 72,
      "red": 8,
    },
    Object {
      "blue": 22,
      "green": 72,
      "red": 36,
    },
    Object {
      "blue": 20,
      "green": 83,
      "red": 40,
    },
    Object {
      "blue": 3,
      "green": 4,
      "red": 218,
    },
    Object {
      "blue": 6,
      "green": 7,
      "red": 186,
    },
    Object {
      "blue": 9,
      "green": 56,
      "red": 8,
    },
    Object {
      "blue": 1,
      "green": 0,
      "red": 0,
    },
    Object {
      "blue": 35,
      "green": 74,
      "red": 113,
    },
    Object {
      "blue": 38,
      "green": 71,
      "red": 104,
    },
    Object {
      "blue": 103,
      "green": 249,
      "red": 250,
    },
    Object {
      "blue": 5,
      "green": 6,
      "red": 199,
    },
    Object {
      "blue": 39,
      "green": 56,
      "red": 68,
    },
    Object {
      "blue": 182,
      "green": 129,
      "red": 12,
    },
    Object {
      "blue": 19,
      "green": 91,
      "red": 50,
    },
    Object {
      "blue": 14,
      "green": 97,
      "red": 18,
    },
    Object {
      "blue": 23,
      "green": 54,
      "red": 86,
    },
    Object {
      "blue": 6,
      "green": 8,
      "red": 169,
    },
    Object {
      "blue": 37,
      "green": 52,
      "red": 56,
    },
    Object {
      "blue": 11,
      "green": 39,
      "red": 11,
    },
    Object {
      "blue": 174,
      "green": 122,
      "red": 7,
    },
    Object {
      "blue": 10,
      "green": 12,
      "red": 105,
    },
    Object {
      "blue": 9,
      "green": 11,
      "red": 133,
    },
    Object {
      "blue": 25,
      "green": 67,
      "red": 108,
    },
    Object {
      "blue": 14,
      "green": 71,
      "red": 35,
    },
    Object {
      "blue": 35,
      "green": 60,
      "red": 85,
    },
    Object {
      "blue": 40,
      "green": 142,
      "red": 77,
    },
    Object {
      "blue": 28,
      "green": 52,
      "red": 57,
    },
    Object {
      "blue": 16,
      "green": 109,
      "red": 54,
    },
    Object {
      "blue": 32,
      "green": 151,
      "red": 63,
    },
    Object {
      "blue": 128,
      "green": 92,
      "red": 11,
    },
    Object {
      "blue": 32,
      "green": 137,
      "red": 62,
    },
    Object {
      "blue": 36,
      "green": 71,
      "red": 89,
    },
    Object {
      "blue": 209,
      "green": 107,
      "red": 209,
    },
    Object {
      "blue": 21,
      "green": 179,
      "red": 219,
    },
    Object {
      "blue": 35,
      "green": 165,
      "red": 198,
    },
    Object {
      "blue": 11,
      "green": 62,
      "red": 113,
    },
    Object {
      "blue": 242,
      "green": 122,
      "red": 242,
    },
    Object {
      "blue": 87,
      "green": 68,
      "red": 10,
    },
    Object {
      "blue": 86,
      "green": 70,
      "red": 21,
    },
    Object {
      "blue": 71,
      "green": 58,
      "red": 9,
    },
    Object {
      "blue": 113,
      "green": 82,
      "red": 12,
    },
    Object {
      "blue": 12,
      "green": 25,
      "red": 40,
    },
    Object {
      "blue": 27,
      "green": 52,
      "red": 38,
    },
    Object {
      "blue": 20,
      "green": 26,
      "red": 55,
    },
    Object {
      "blue": 22,
      "green": 24,
      "red": 85,
    },
    Object {
      "blue": 9,
      "green": 11,
      "red": 119,
    },
    Object {
      "blue": 26,
      "green": 45,
      "red": 12,
    },
    Object {
      "blue": 90,
      "green": 242,
      "red": 245,
    },
    Object {
      "blue": 54,
      "green": 40,
      "red": 53,
    },
    Object {
      "blue": 22,
      "green": 187,
      "red": 229,
    },
    Object {
      "blue": 51,
      "green": 66,
      "red": 23,
    },
    Object {
      "blue": 39,
      "green": 53,
      "red": 12,
    },
    Object {
      "blue": 71,
      "green": 140,
      "red": 138,
    },
    Object {
      "blue": 65,
      "green": 139,
      "red": 123,
    },
    Object {
      "blue": 37,
      "green": 43,
      "red": 53,
    },
    Object {
      "blue": 75,
      "green": 44,
      "red": 75,
    },
    Object {
      "blue": 28,
      "green": 34,
      "red": 69,
    },
    Object {
      "blue": 19,
      "green": 24,
      "red": 72,
    },
    Object {
      "blue": 10,
      "green": 23,
      "red": 53,
    },
    Object {
      "blue": 83,
      "green": 72,
      "red": 46,
    },
    Object {
      "blue": 67,
      "green": 65,
      "red": 18,
    },
    Object {
      "blue": 83,
      "green": 170,
      "red": 175,
    },
    Object {
      "blue": 12,
      "green": 57,
      "red": 20,
    },
    Object {
      "blue": 40,
      "green": 81,
      "red": 91,
    },
    Object {
      "blue": 39,
      "green": 129,
      "red": 152,
    },
    Object {
      "blue": 105,
      "green": 57,
      "red": 105,
    },
    Object {
      "blue": 123,
      "green": 65,
      "red": 123,
    },
    Object {
      "blue": 57,
      "green": 139,
      "red": 109,
    },
    Object {
      "blue": 41,
      "green": 42,
      "red": 69,
    },
    Object {
      "blue": 12,
      "green": 17,
      "red": 86,
    },
    Object {
      "blue": 15,
      "green": 20,
      "red": 96,
    },
    Object {
      "blue": 14,
      "green": 85,
      "red": 41,
    },
    Object {
      "blue": 17,
      "green": 17,
      "red": 124,
    },
    Object {
      "blue": 51,
      "green": 74,
      "red": 51,
    },
    Object {
      "blue": 97,
      "green": 221,
      "red": 229,
    },
    Object {
      "blue": 55,
      "green": 52,
      "red": 10,
    },
    Object {
      "blue": 53,
      "green": 119,
      "red": 101,
    },
    Object {
      "blue": 223,
      "green": 224,
      "red": 223,
    },
    Object {
      "blue": 42,
      "green": 110,
      "red": 128,
    },
    Object {
      "blue": 147,
      "green": 77,
      "red": 147,
    },
    Object {
      "blue": 33,
      "green": 186,
      "red": 224,
    },
    Object {
      "blue": 35,
      "green": 150,
      "red": 180,
    },
    Object {
      "blue": 18,
      "green": 193,
      "red": 237,
    },
    Object {
      "blue": 27,
      "green": 85,
      "red": 100,
    },
    Object {
      "blue": 80,
      "green": 156,
      "red": 162,
    },
    Object {
      "blue": 28,
      "green": 64,
      "red": 74,
    },
    Object {
      "blue": 7,
      "green": 192,
      "red": 239,
    },
    Object {
      "blue": 41,
      "green": 80,
      "red": 119,
    },
    Object {
      "blue": 29,
      "green": 53,
      "red": 71,
    },
    Object {
      "blue": 12,
      "green": 19,
      "red": 76,
    },
    Object {
      "blue": 48,
      "green": 47,
      "red": 79,
    },
    Object {
      "blue": 51,
      "green": 51,
      "red": 71,
    },
    Object {
      "blue": 20,
      "green": 21,
      "red": 102,
    },
    Object {
      "blue": 59,
      "green": 122,
      "red": 117,
    },
    Object {
      "blue": 45,
      "green": 113,
      "red": 81,
    },
    Object {
      "blue": 175,
      "green": 89,
      "red": 174,
    },
    Object {
      "blue": 10,
      "green": 38,
      "red": 56,
    },
    Object {
      "blue": 48,
      "green": 71,
      "red": 77,
    },
    Object {
      "blue": 70,
      "green": 60,
      "red": 40,
    },
    Object {
      "blue": 69,
      "green": 74,
      "red": 85,
    },
    Object {
      "blue": 189,
      "green": 97,
      "red": 189,
    },
    Object {
      "blue": 10,
      "green": 43,
      "red": 22,
    },
    Object {
      "blue": 77,
      "green": 90,
      "red": 82,
    },
    Object {
      "blue": 33,
      "green": 115,
      "red": 58,
    },
    Object {
      "blue": 25,
      "green": 74,
      "red": 87,
    },
    Object {
      "blue": 32,
      "green": 174,
      "red": 209,
    },
    Object {
      "blue": 25,
      "green": 130,
      "red": 156,
    },
    Object {
      "blue": 54,
      "green": 135,
      "red": 155,
    },
    Object {
      "blue": 51,
      "green": 155,
      "red": 180,
    },
    Object {
      "blue": 31,
      "green": 113,
      "red": 133,
    },
    Object {
      "blue": 30,
      "green": 95,
      "red": 112,
    },
    Object {
      "blue": 35,
      "green": 117,
      "red": 137,
    },
    Object {
      "blue": 27,
      "green": 138,
      "red": 166,
    },
    Object {
      "blue": 25,
      "green": 163,
      "red": 198,
    },
    Object {
      "blue": 38,
      "green": 139,
      "red": 165,
    },
    Object {
      "blue": 39,
      "green": 147,
      "red": 174,
    },
    Object {
      "blue": 38,
      "green": 85,
      "red": 97,
    },
    Object {
      "blue": 41,
      "green": 160,
      "red": 190,
    },
    Object {
      "blue": 91,
      "green": 219,
      "red": 226,
    },
    Object {
      "blue": 38,
      "green": 37,
      "red": 87,
    },
    Object {
      "blue": 102,
      "green": 81,
      "red": 39,
    },
    Object {
      "blue": 84,
      "green": 185,
      "red": 193,
    },
    Object {
      "blue": 50,
      "green": 163,
      "red": 99,
    },
    Object {
      "blue": 96,
      "green": 190,
      "red": 201,
    },
    Object {
      "blue": 122,
      "green": 100,
      "red": 25,
    },
    Object {
      "blue": 160,
      "green": 160,
      "red": 158,
    },
    Object {
      "blue": 85,
      "green": 100,
      "red": 92,
    },
    Object {
      "blue": 48,
      "green": 88,
      "red": 67,
    },
    Object {
      "blue": 54,
      "green": 69,
      "red": 84,
    },
    Object {
      "blue": 51,
      "green": 109,
      "red": 96,
    },
    Object {
      "blue": 78,
      "green": 124,
      "red": 135,
    },
    Object {
      "blue": 81,
      "green": 123,
      "red": 81,
    },
    Object {
      "blue": 98,
      "green": 204,
      "red": 211,
    },
    Object {
      "blue": 77,
      "green": 86,
      "red": 77,
    },
    Object {
      "blue": 64,
      "green": 145,
      "red": 63,
    },
    Object {
      "blue": 51,
      "green": 164,
      "red": 193,
    },
    Object {
      "blue": 32,
      "green": 69,
      "red": 78,
    },
    Object {
      "blue": 8,
      "green": 189,
      "red": 234,
    },
    Object {
      "blue": 29,
      "green": 99,
      "red": 117,
    },
    Object {
      "blue": 15,
      "green": 53,
      "red": 62,
    },
    Object {
      "blue": 51,
      "green": 142,
      "red": 165,
    },
    Object {
      "blue": 39,
      "green": 97,
      "red": 111,
    },
    Object {
      "blue": 41,
      "green": 105,
      "red": 120,
    },
    Object {
      "blue": 50,
      "green": 45,
      "red": 11,
    },
    Object {
      "blue": 34,
      "green": 38,
      "red": 13,
    },
    Object {
      "blue": 54,
      "green": 59,
      "red": 19,
    },
    Object {
      "blue": 49,
      "green": 40,
      "red": 17,
    },
    Object {
      "blue": 90,
      "green": 203,
      "red": 211,
    },
    Object {
      "blue": 105,
      "green": 114,
      "red": 108,
    },
    Object {
      "blue": 97,
      "green": 180,
      "red": 185,
    },
    Object {
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "blue": 0,
      "green": 0,
      "red": 0,
    },
  ],
  "header": Object {
    "magic": "GIF",
    "version": "89a",
  },
  "logicalScreen": Object {
    "backgroundColorIndex": 1,
    "colorResolution": 7,
    "globalColorTable": 1,
    "globalColorTableSize": 7,
    "imageHeight": 189,
    "imageWidth": 150,
    "pixelAspectRatio": 0,
    "sort": 1,
  },
}
`;
